### DISCUSSION TITLE
8048S070C: kernel panic

### DISCUSSION THREAD

**Discussion by lunard** [rzeldent/esp32-smartdisplay#164] [Q&A] [answered]
Hello to all and first thanks for this library!!
I'm trying to run an easy demo, but my device (8048S070C) always crashes when I execute the function  smartdisplay_init();

Maybe someone can help me ?
Here my config and code:

platform.ini
```
[platformio]
default_envs = esp32-8048S070C

[env]
platform = espressif32
framework = arduino
lib_deps = 
	rzeldent/esp32_smartdisplay

monitor_speed = 115200
monitor_rts = 0
monitor_dtr = 0
monitor_filters = esp32_exception_decoder
board_build.partitions = min_spiffs.csv
build_flags = 
	-Ofast
	-Wall
    #-D CORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_DEBUG
    #-D CORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_INFO
	-D CORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_VERBOSE
	# -DLV_CONF_PATH=${PROJECT_INCLUDE_DIR}/lv_conf.h
	-D LV_CONF_INCLUDE_SIMPLE  

[env:esp32-8048S070C]
board = esp32-8048S070C
```

main.cpp

```#include <Arduino.h>
#include "UI/ui.h"
#include <esp32_smartdisplay.h>

void DisplayDebugInfo()
{
  Serial.setDebugOutput(true);
  log_i("CPU: %s rev%d, CPU Freq: %d Mhz, %d core(s)", ESP.getChipModel(), ESP.getChipRevision(), getCpuFrequencyMhz(), ESP.getChipCores());
  log_i("Free heap: %d bytes", ESP.getFreeHeap());
  log_i("Free PSRAM: %d bytes", ESP.getPsramSize());
  log_i("SDK version: %s", ESP.getSdkVersion());
}

void setup()
{
  Serial.begin(115200);
  DisplayDebugInfo();

  // smartdisplay_init(); <-- CRASH !!

  __attribute__((unused)) auto disp = lv_disp_get_default();
  // lv_disp_set_rotation(disp, LV_DISP_ROT_90);
  // lv_disp_set_rotation(disp, LV_DISP_ROT_180);
  // lv_disp_set_rotation(disp, LV_DISP_ROT_270);

  // Start the UI
  // ui_init();
}

void loop()
{
  // lv_timer_handler();
}
```

Here the stack trace

Backtrace: 0x40377eb2:0x3fceb9a0 0x4037b6e5:0x3fceb9c0 0x40381605:0x3fceb9e0 0x4037b6db:0x3fceba60 0x42025fa8:0x3fceba80 0x42025c6a:0x3fcebc00 0x420022f4:0x3fcebc20 0x42029242:0x3fcebc50

  #0  0x40377eb2:0x3fceb9a0 in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:408
  #1  0x4037b6e5:0x3fceb9c0 in esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:137      
  #2  0x40381605:0x3fceb9e0 in abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/abort.c:46
  #3  0x4037b6db:0x3fceba60 in _esp_error_check_failed at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_err.c:43   
  #4  0x42025fa8:0x3fceba80 in lvgl_lcd_init at .pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/lvgl_panel_st7262_par.c:69 (discriminator 1)  
  #5  0x42025c6a:0x3fcebc00 in smartdisplay_init at .pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c:219
  #6  0x420022f4:0x3fcebc20 in setup() at src/main.cpp:19
  #7  0x42029242:0x3fcebc50 in loopTask(void*) at C:/Users/lucan/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:42

**Comment by xaljox** [comment]
I have a similar issue with the 8048S043C board. I created a new project with the supplied json board file. Followed the tutorial, edited the lc_conv.h and platformIo.ini as indicated. Have a simple main.cpp which crashes in the smartdisplay_int(). 
A more complex example using LVGL with an GFX library works, but I wanted to see if this was a better option. But with a crash it isn't possible to see if it is any better.

**Comment by xaljox** [comment]
This is my trace stuf:
Rebooting...
␀�ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x4204d29e
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x4bc
load:0x403c9700,len:0xbd8
load:0x403cc700,len:0x2a0c
entry 0x403c98d0
[   134][I][esp32-hal-psram.c:96] psramInit(): PSRAM enabled
[   145][D][esp32_smartdisplay.c:173] smartdisplay_init(): smartdisplay_init
[Info]  (0.152, +152)    lv_init: begin         (in lv_obj.c line #106)
[   153][V][lvgl_panel_st7262_par.c:23] lvgl_lcd_init(): drv:0x00000006
[   164][D][lvgl_panel_st7262_par.c:65] lvgl_lcd_init(): rgb_panel_config: clk_src:0, 
            timings:{pclk_hz:12500000, h_res:800, v_res:480, hsync_pulse_width:4, 
            hsync_back_porch:8, hsync_front_porch:8, vsync_pulse_width:4, vsync_back_porch:8, 
            vsync_front_porch:8, flags:{hsync_idle_low:0, vsync_idle_low:0, de_idle_high:0, 
            pclk_active_neg:1, pclk_idle_high:0}}, data_width:16, sram_trans_align:4, 
            psram_trans_align:64, hsync_gpio_num:39, vsync_gpio_num:41, de_gpio_num:40, 
            pclk_gpio_num:42, data_gpio_nums:[45,48,47,21,14,5,6,7,15,16,4,8,3,46,9,1,], 
            disp_gpio_num:-1, on_frame_trans_done:0x420277e4, user_ctx:0x3fca44b0, 
            flags:{disp_active_low:0, relax_on_idle:0, fb_in_psram:1}
[   223][D][lvgl_panel_st7262_par.c:66] lvgl_lcd_init(): refresh rate: 0 Hz
ESP_ERROR_CHECK failed: esp_err_t 0x106 (ESP_ERR_NOT_SUPPORTED) at 0x4037b5d8
file: ".pio/libdeps/esp32-8048S043C/esp32_smartdisplay/src/lvgl_panel_st7262_par.c" line 76
func: lvgl_lcd_init
expression: esp_lcd_panel_swap_xy(panel_handle, DISPLAY_SWAP_XY)

abort() was called at PC 0x4037b5db on core 1


Backtrace: 0x40377e12:0x3fceb9b0 0x4037b5e5:0x3fceb9d0 0x40381559:0x3fceb9f0 
           0x4037b5db:0x3fceba70 0x42027ad4:0x3fceba90 0x420277a7:0x3fcebc10 
           0x420026fb:0x3fcebc30 0x42029782:0x3fcebc50

**Comment by xaljox** [comment, marked as answer]
Ik think I found the issue in my case the code was:

#ifdef DISPLAY_SWAP_XY
    ESP_ERROR_CHECK(esp_lcd_panel_swap_xy(panel_handle, DISPLAY_SWAP_XY));
#endif
#if defined(DISPLAY_MIRROR_X) || defined(DISPLAY_MIRROR_Y)    
    ESP_ERROR_CHECK(esp_lcd_panel_mirror(panel_handle, DISPLAY_MIRROR_X, DISPLAY_MIRROR_Y));
#endif    
#if defined(DISPLAY_GAP_X) || defined(DISPLAY_GAP_Y)
    ESP_ERROR_CHECK(esp_lcd_panel_set_gap(panel_handle, DISPLAY_GAP_X, DISPLAY_GAP_Y));
#endif

and DISPLAY_SWAP_XY is defined as build flag in the json file. It is set at false, but still it is a defined define. Thus either those defines must be deleted from the json file, or the code must be:
#if(DISPLAY_SWAP_XY)
    ESP_ERROR_CHECK(esp_lcd_panel_swap_xy(panel_handle, DISPLAY_SWAP_XY));
#endif
#if(DISPLAY_MIRROR_X || DISPLAY_MIRROR_Y)    
    ESP_ERROR_CHECK(esp_lcd_panel_mirror(panel_handle, DISPLAY_MIRROR_X, DISPLAY_MIRROR_Y));
#endif    
#if defined(DISPLAY_GAP_X) || defined(DISPLAY_GAP_Y)
    ESP_ERROR_CHECK(esp_lcd_panel_set_gap(panel_handle, DISPLAY_GAP_X, DISPLAY_GAP_Y));
#endif

**Reply by rzeldent** [reply]
Think you're so right!

Made an update in develop branch to use the statements below.
This because would like to keep the variables in case software rotation is present.

Think this works!

#if (DISPLAY_SWAP_XY)
#if (DISPLAY_MIRROR_X || DISPLAY_MIRROR_Y) 
#if (DISPLAY_GAP_X || DISPLAY_GAP_Y)

**Reply by rzeldent** [reply]
Could you check if the issue is resolved with this change?

You can check this by using in the platformio in the lib_deps the develop branch:

lib_deps = https://github.com/rzeldent/esp32-smartdisplay#develop

Thanks!

**Comment by xaljox** [comment]
I put the line in as suggested, when I saved the platform.ini file it updated the library. My edits in the file 'lvl_panel_st7262_par.c' were gone and I got back the old wrong file. I cmd-clicked the line in the platform.ini file to open the repository in the browser and there was also the wrong file. See also the screenshot.
Thus somehow I do not get the updates you've made.

￼

**Comment by xaljox** [comment]
I've also spotted an error I think in the 'framework-arduinoespressif32/cores/esp32/esp32-hal-ledc.c' file.
The code is:
void ledcWrite(uint8_t chan, uint32_t duty)
{
    if(chan >= LEDC_CHANNELS){
        return;
    }
    uint8_t group=(chan/8), channel=(chan%8);

    //Fixing if all bits in resolution is set = LEDC FULL ON
    uint32_t max_duty = (1 << channels_resolution[chan]) - 1;

    if((duty == max_duty) && (max_duty != 1)){
        duty = max_duty + 1;
    }

    ledc_set_duty(group, channel, duty);
    ledc_update_duty(group, channel);
}

But I think the red coloured lines are wrong, it should be:
    if((duty > max_duty) && (max_duty != 1)){
        duty = max_duty;
    }

Any idea were to put this update, I'm not an Github guru. I do not even know the official place the code comes from.

With the original code when the brightness is set at full brightness e.g. 255 with 8 bits resolution, then duty is equal to max_duty and then duty is set to 256 which gives en wire error and the serial output is spawn with error messages.
